### PR TITLE
Update ddp config to improve ESM-2 15B MFU

### PR DESCRIPTION
**Pull request by sichu2023**
Update DDP config to speed up ESM-2 15B pretraining. Turn off `grad_reduce_in_fp32` in mixed precision plugin (default is True) to reduce memory consumption and `overlap_grad_reduce` and `average_in_collective` to improve performance.

Pause `overlap_param_gather=True` to wait for NeMo's fix.

### CODE REVIEW

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py [hunk 1]
```diff
@@ -163,10 +164,17 @@ def main(
     strategy = nl.MegatronStrategy(
         tensor_model_parallel_size=tensor_model_parallel_size,
         pipeline_model_parallel_size=pipeline_model_parallel_size,
-        ddp="megatron",
+        pipeline_dtype=get_autocast_dtype(precision),
+        ddp=DistributedDataParallelConfig(
+            check_for_nan_in_grad=True,
+            overlap_grad_reduce=True,
+            overlap_param_gather=True,
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:171]: Does this make us possibly exposed to the loss spike NeMo bug where they saw issues with overlap_param_gather?

Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:171]: Good call. Do they have a PR merged?

Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py]: Took it off for the moment.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py [hunk 2]
```diff
@@ -163,10 +164,17 @@ def main(
...
         tensor_model_parallel_size=tensor_model_parallel_size,
         pipeline_model_parallel_size=pipeline_model_parallel_size,
-        ddp="megatron",
+        pipeline_dtype=get_autocast_dtype(precision),
+        ddp=DistributedDataParallelConfig(
+            check_for_nan_in_grad=True,
+            overlap_grad_reduce=True,
+            overlap_param_gather=False,  # TODO waiting for a NeMo fix
+            average_in_collective=True,
+            use_distributed_optimizer=True,
```
Comment (jstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:173]: maybe make grad reduce in fp32 an option or set to True? True seems to be the default in nearly all of the nemo recipes. Also note that I'm not clear on whether the ddp grad reduce in fp32 or the megatron mixed precision takes precedent.

File: sub-packages/bionemo-llm/src/bionemo/llm/train.py [hunk 3]
```diff
@@ -107,10 +109,17 @@ def setup_trainer(
     strategy = nl.MegatronStrategy(
         tensor_model_parallel_size=parallel_config.tensor_model_parallel_size,
         pipeline_model_parallel_size=parallel_config.pipeline_model_parallel_size,
-        ddp="megatron",
+        pipeline_dtype=get_autocast_dtype(training_config.precision),
+        ddp=DistributedDataParallelConfig(
+            check_for_nan_in_grad=True,
+            overlap_grad_reduce=True,
+            overlap_param_gather=True,
+            average_in_collective=True,
```
Comment (jstjohn) [inline sub-packages/bionemo-llm/src/bionemo/llm/train.py:117]: same comments here related to fp32 grad reduce

**Review by jstjohn — APPROVED** [review]

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py (hunk 2 above)
Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:181]: The problem is `grad_reduce_fp32` hurts perf.

Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:181]: Added in cli but disabled by default.

**Review by pstjohn — APPROVED** [review]
Does this change any of the current defaults? Might be good to call those out in the PR body if so

**Comment by sichu2023** [comment]
> Does this change any of the current defaults? Might be good to call those out in the PR body if so

Thanks. Just added.

**Comment by sichu2023** [comment]
With the updated ddp config, test run on 650m pretraining results in 0.01 lower training loss and 0.01 higher validation loss, both discrepancy smaller than 1% of of the target value (~2.7).

W B Chart 12_30_2024, 9_28_23 AM
W B Chart 12_30_2024, 9_28_42 AM

**Comment by sichu2023** [comment]
Dependent on doc fix https://github.com/NVIDIA/bionemo-framework/pull/559
